feat(iota-core,iota-types,iota-framework,iota-node): Implement a mechanism that excludes inactive validators

piotrm50 opened the pull request:
# Description of change

This PR implements a comprehensive capability-based committee selection system for validators, enabling dynamic committee membership based on protocol version support and validator readiness. The implementation spans three key components: capability advertisement infrastructure, enhanced committee selection logic, and automated capability notification.

This is a PR to merge a feature branch to develop. Previous PRs that form this one have been reviewed by the respective code owners. 

## Links to any relevant issues

Resolves [#7814](https://github.com/iotaledger/iota/issues/7814)

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

- [x] Protocol: 
   -  Introduced `AdvanceEpochV3` which allows committee selection from  an eligible validators vector. By default, all active validators are eligible for selection. This feature is enabled on all networks.
  - Added `select_committee_from_eligible_validators` feature flag to further restrict the eligible validators vector to only those active validators that support the next epoch's protocol version and have correctly signaled their support. This feature is currently enabled only on devnet.
- [x] Nodes (Validators and Full nodes): Added gRPC endpoint allowing non-committee active validators to forward signed `AuthorityCapabilities` messages to committee members. Non-committee active validators now automatically send `SignedAuthorityCapabilities` messages to signal their eligibility for committee selection upon node startup or epoch reconfiguration. This functionality is enabled on all networks.
- [ ] Indexer:
- [x] JSON-RPC: Extend `IotaChangeEpochV2` type with an optional field `eligible_active_validators: Option<Vec<u64>>` 
- [x] GraphQL: Extend `ChangeEpochTransactionV2` type with an optional field `eligible_active_validators: Option<Vec<u64>>` 
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

piotrm50 committed:
feat(iota-core,iota-types): Add AuthorityCapabilities endpoint  (#8101)

# Description of change

This PR implements the authority capabilities advertisement endpoint for
non-committee validators, enabling them to notify committee validators
about their capabilities and supported protocol versions. This feature
lays the groundwork for improved validator coordination and protocol
version management across the network that will be implemented in
subsequent PRs.

Key Features

- New message format for non-committee active validators to advertise
their capabilities.
- Signature Verification: Enhanced signature verifier to validate
capability notifications from authorized non-committee active validators
- gRPC Endpoint: New handle_capability_notification_v1 endpoint for
receiving capability advertisements
- Epoch State V2: Extended epoch start state to include non-committee
active validator information
- Comprehensive Testing: Added unit tests to verify proper rejection of
unauthorized capability notifications

## Links to any relevant issues

Resolves #7992

## How the change has been tested

- Unit tests verify proper rejection of capability notifications from
unauthorized authorities
- Further tests will be implemented in subsequent PRs as end-to-end
tests.

Make sure to provide instructions for the maintainer as well as any
relevant configurations.

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items
that are not.

- [x] I have followed the [contribution
guidelines](../../CONTRIBUTING.md) for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### Release Notes

Check each box that your changes affect. If none of the boxes relate to
your changes, release notes aren't required.

For each box you select, include information after the relevant heading
that describes the impact of your changes that a user might notice and
any actions they must take to implement updates.

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): Adds gRPC endpoint to allow
non-committee active validators to forward their signed
AuthorityCapabilities message to committee members. This will enable
selecting new committee only among validators supporting new protocol
version when advancing epoch. This endpoint is only enabled on the
devnet.
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

piotrm50 committed:
feat(iota-core,iota-types,iota-framework): Process AuthorityCapabilities from non-committee validators (#8127)

# Description of change

This PR updates the way committee selection is done by limiting eligible
active validators only to those that support the ProtocolVersion of the
upcoming epoch. This involves introducing ChangeEpochV3 which contains
indices of the active validators supporting upcoming ProtocolVersion.
There is a fallback to select among all active validators if
eligible_active_validators set does not contain a quorum of current
committee.

## Links to any relevant issues

Resolves #7993 #7994 

## How the change has been tested

Full suite of tests and CI fixes will be implemented in the following
PR.

- [ ] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [x] Protocol: Improved committee selection where committee is selected
only from the list of vector of eligible active validators. This will be
enabled on all networks with the same protocol version. By default all
active validators are considered eligible, however when
`track_non_committee_eligible_validators` feature flag is enabled, then
the list of eligible validators is composed of committee and
non-committee active validators that have issued AuthorityCapabilities
to signal their readiness to become a committee member as well as
announce that they support the next epoch's protocol version.
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

---------

Co-authored-by: Konstantinos Demartinos <konstantinos.demartinos@iota.org>

piotrm50 committed:
feat(iota-core,iota-node): Capability notification sending for non-committee validators (#8296)

# Description of change

This PR introduces ability to send CapabilityNotification by
non-committee validators to be taken into consideration during the next
committee selection process.

## Links to any relevant issues

Part of #

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): Ability to send
CapabilityNotification by non-committee active validators to be taken
into consideration during the next committee selection process.
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

piotrm50 committed:
feat(iota-core): Include digest when creating `eligible_validators` vector (#8776)

# Description of change

Include digest when creating `eligible_validators` vector. 
Check that required protocol parameters are set when retrieving
`select_committee_from_eligible_validators` param.

## Links to any relevant issues

Partly covers #7814  

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

miker83z commented on crates/iota-framework-snapshot/manifest.json:
```diff
@@ -338,5 +338,30 @@
     ]
+  },
+  "13": {
+    "git_revision": "e0f4712359e5f92d69982a54bbce05f2954319ca",
```
Since this is a feature branch and I guess it won't be squashed before merging to develop. Then, please create a dedicated commit in which you update this revision to `e0a595ae78233a7aca2f1f6c92b0872382a06139`. 
Instead, if you intend to squash, then we need a commit with an update once merged to develop.

piotrm50 committed:
feat(iota-protocol-config,iota-core): Add feature flag for selecting committee supporting next epoch version (#8788)

# Description of change

Add a separate feature flag for selecting committee supporting next
epoch version instead of functionality being covered by
`track_non_committee_eligible_validators` flag that controls
sending/processing of AuthorityCapabilities notifications from
non-committee active validators.

## Links to any relevant issues

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

piotrm50 committed:
fix(sdk): Regenerate autogenerated GraphQL files.  (#8793)

# Description of change

Regenerate autogenerated GraphQL files. 

## Links to any relevant issues

Partly resolves https://github.com/iotaledger/iota/issues/7814

## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

---------

Co-authored-by: Mario Sarcevic <mario.sarcevic@iota.org>

piotrm50 committed:
feat(iota-framework): Update git revision in manifest.json (#8806)

# Description of change

Update git revision in manifest.json

## Links to any relevant issues



## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes
